### PR TITLE
Clips pitch and speed

### PR DESCRIPTION
**Pull request by Eism**
Resolves: #6946
Resolves: #6950

### CODE REVIEW

File: src/projectscene/internal/projectsceneactionscontroller.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@ using namespace muse::actions;
 static const ActionCode VERTICAL_RULERS_CODE("toggle-vertical-rulers");
 static const ActionCode MINUTES_SECONDS_RULER("minutes-seconds-ruler");
 static const ActionCode BEATS_MEASURES_RULER("beats-measures-ruler");
+static const ActionCode PITCH_AND_SPEED_CODE("clip-pitch-speed");
```
Comment (igorkorsukov) [inline src/projectscene/internal/projectsceneactionscontroller.cpp:15]: adding these constants arose for historical reasons, there is usually no point in them, it is even a little inconvenient to compare the action names with the handler

File: src/projectscene/internal/projectsceneactionscontroller.cpp [hunk 2]
```diff
@@ -66,6 +68,21 @@ void ProjectSceneActionsController::pinnedPlayHead()
     NOT_IMPLEMENTED;
 }
 
+void ProjectSceneActionsController::openClipPitchAndSpeedEdit(const ActionData& args)
+{
+    trackedit::ClipKey clipKey = args.arg<trackedit::ClipKey>(0);
```
Comment (igorkorsukov) [inline src/projectscene/internal/projectsceneactionscontroller.cpp:73]: It's better to add a check for the count of arguments
```
IF_ASSERT_FAILED(args.count() == 1) {
        return;
}
```

File: src/trackedit/internal/trackeditactionscontroller.cpp [hunk 3]
```diff
@@ -578,6 +582,18 @@ void TrackeditActionsController::silenceAudioSelection()
     pushProjectHistoryTrackSilenceState(selectedStartTime, selectedEndTime);
 }
 
+void TrackeditActionsController::renderClipPitchAndSpeed(const muse::actions::ActionData& args)
+{
+    trackedit::ClipKey clipKey = args.arg<trackedit::ClipKey>(0);
```
Comment (igorkorsukov) [inline src/trackedit/internal/trackeditactionscontroller.cpp:587]: same https://github.com/audacity/audacity/pull/7600/files#r1810627325

File: src/trackedit/internal/trackeditactionscontroller.cpp [hunk 4]
```diff
@@ -578,6 +582,18 @@ void TrackeditActionsController::silenceAudioSelection()
...
+    if (!clipKey.isValid()) {
+        return;
+    }
+
+    trackeditInteraction()->renderClipPitchAndSpeed(clipKey);
+
+    pushProjectHistoryRenderClipStretchingState();
```
Comment (igorkorsukov) [inline src/trackedit/internal/trackeditactionscontroller.cpp:594]: This should be part of the logic of the trackeditInteraction()->renderClipPitchAndSpeed method, i.e. if we call trackeditInteraction()->renderClipPitchAndSpeed from some place, then the logic associated with history should also be executed

File: src/projectscene/view/clipsview/pitchandspeedchangemodel.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+int PitchAndSpeedChangeModel::speedPercentage() const
+{
+    return std::round(100.0 / m_clip.speed);
+}
+
+void PitchAndSpeedChangeModel::setSpeedPercentage(int speedPercentage)
+{
+    double speed = 100.0 / double(speedPercentage);
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/projectscene/view/clipsview/pitchandspeedchangemodel.cpp:60]: do we even need a cast? speedPercentage will be implicitly promoted to double in any case.

File: src/trackedit/internal/au3/au3interaction.cpp [hunk 6]
```diff
@@ -304,6 +305,90 @@ bool Au3Interaction::changeClipTitle(const trackedit::ClipKey& clipKey, const mu
...
+    std::shared_ptr<WaveClip> clip = DomAccessor::findWaveClip(waveTrack, clipKey.clipId);
+    IF_ASSERT_FAILED(clip) {
+        return false;
+    }
+
+    waveTrack->ApplyPitchAndSpeed({ { clip->GetPlayStartTime(), clip->GetPlayEndTime() } }, {});
+    LOGD() << "apply pitch and speed for clip: " << clipKey.clipId << ", track: " << clipKey.trackId;
+
+    trackedit::ITrackeditProjectPtr prj = globalContext()->currentTrackeditProject();
+    prj->onTrackChanged(DomConverter::track(waveTrack)); //! todo: replase with onClipChanged
```
[the remark this answers is not in the document]
Comment (Eism) [reply src/trackedit/internal/au3/au3interaction.cpp:397]: as I see, the method takes in the start and end time of the clip and inside only this clip changes, visually only the clip also changes

File: src/projectscene/internal/projectsceneactionscontroller.cpp (hunk 1 above)
Comment (Eism) [reply src/projectscene/internal/projectsceneactionscontroller.cpp]: renamed

File: src/projectscene/qml/Audacity/ProjectScene/clipsview/ClipItem.qml [hunk 7]
```diff
@@ -243,22 +252,51 @@ Rectangle {
                 id: buttonsRow
 
                 anchors.right: parent.right
-                anchors.rightMargin: 4
+                anchors.rightMargin: root.rightVisibleMargin + 4
```
Comment (grliszas14) [inline src/projectscene/qml/Audacity/ProjectScene/clipsview/ClipItem.qml:255]: I'm just gonna leave a comment here so we don't forget to clarify what should happen with hiding icons beyond right edge of the screen.
Not mentioning anybody as discussion is handled somewhere else I see

**Review by grliszas14 — CHANGES_REQUESTED** [review]

**Review by igorkorsukov — APPROVED** [review]

File: src/projectscene/qml/Audacity/ProjectScene/clipsview/ClipItem.qml (hunk 7 above)
Comment (Eism) [reply src/projectscene/qml/Audacity/ProjectScene/clipsview/ClipItem.qml]: dropped the commit with this feature

**Review by grliszas14 — APPROVED** [review]

**Pull request merged**